### PR TITLE
HTML-755: Conversion from client to server timezone.

### PR DESCRIPTION
**Pull request by icrc-jfigueiredo**
[DRAFT] HTML-755 Convert from client timezone to server timezone before saving on DB.
This only works if the`<encounterDate>`  tag have ` showTime=true `

### CODE REVIEW

File: api/src/main/java/org/openmrs/module/htmlformentry/widget/TimeWidget.java [hunk 1]
```diff
@@ -49,8 +49,14 @@ public String generateHtml(FormEntryContext context) {
...
 			String toPrint = "";
 			if (initialValue != null) {
+				StringBuilder sb = new StringBuilder();
 				toPrint = timeFormat().format(initialValue);
-				return WidgetFactory.displayValue(toPrint);
+				//Get server timezone
+				SimpleDateFormat dateFormatYouWant = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ssZ");
+				String newDate = dateFormatYouWant.format(initialValue);
+				String timezone = newDate.substring(newDate.length()-5);
+				sb.append("<span class=\"value\" tz=\"").append(timezone).append("\">").append(toPrint).append("</span>");
```
Comment (mks-d) [inline api/src/main/java/org/openmrs/module/htmlformentry/widget/TimeWidget.java:58]: This leaves a possibility for the frontend to convert back to the client timezone in VIEW mode, correct?

Btw you can simply do:
```java
String timezone = new SimpleDateFormat("Z").format(initialValue);
```

File: api/src/main/java/org/openmrs/module/htmlformentry/widget/TimeWidget.java [hunk 2]
```diff
@@ -124,6 +130,9 @@ public String generateHtml(FormEntryContext context) {
 					}
 					sb.append("</select>");
 				}
+				sb.append("<input type=\"hidden\" class=\"hfe-timeZone\" name=\"").append(context.getFieldName(this))
+						.append("timeZone").append("\">");
+				sb.append("</input>");
```
Comment (mks-d) [inline api/src/main/java/org/openmrs/module/htmlformentry/widget/TimeWidget.java:135]: And this creates a possibility for the frontend to submit the timezone in ENTER and EDIT mode, correct?

File: api/src/main/java/org/openmrs/module/htmlformentry/widget/TimeWidget.java [hunk 3]
```diff
@@ -134,6 +143,18 @@ public String generateHtml(FormEntryContext context) {
...
+	public String getTimeZone(FormEntryContext context, HttpServletRequest request) {
+		try {
+			String tz = (String) HtmlFormEntryUtil.getParameterAsType(request, context.getFieldName(this) + "timeZone",
+					String.class);
+			return tz;
+		}
+		catch (Exception ex) {
+			throw new IllegalArgumentException("Illegal value");
+		}
+	}
```
Comment (mks-d) [inline api/src/main/java/org/openmrs/module/htmlformentry/widget/TimeWidget.java:156]: Change this to slightly simplified and renamed:
```java
/**
 * @return The timezone string info that was submitted as part of the time submission. 
 */
public String getSubmittedTimeZone(FormEntryContext context, HttpServletRequest request) {
    return (String) HtmlFormEntryUtil.getParameterAsType(request, context.getFieldName(this) + "timeZone", String.class);
}
```

File: api/src/test/java/org/openmrs/module/htmlformentry/widget/TimeWidgetTest.java [hunk 4]
```diff
@@ -60,7 +61,7 @@ protected void testTimeFormat(TimeWidget widget, String expectedValue) {
 		when(formEntryContext.getMode()).thenReturn(FormEntryContext.Mode.VIEW);
 		String html = widget.generateHtml(formEntryContext);
 		String expected = "<span class=\"value\">" + expectedValue + "</span>";
-		assertEquals(expected, html);
+		assertTrue(StringUtils.containsIgnoreCase(html, expected));
```
Comment (mks-d) [inline api/src/test/java/org/openmrs/module/htmlformentry/widget/TimeWidgetTest.java:64]: I don't think you need to ignore case anymore.

Comment (mks-d) [reply api/src/test/java/org/openmrs/module/htmlformentry/widget/TimeWidgetTest.java:64]: We could do this:
```java
protected void testTimeFormat(TimeWidget widget, String expectedValue) {
    FormEntryContext formEntryContext = mock(FormEntryContext.class);
    when(formEntryContext.getMode()).thenReturn(FormEntryContext.Mode.VIEW);
    String html = widget.generateHtml(formEntryContext);
    String timezone = new SimpleDateFormat("Z").format(new Date());
    String expected = "<span class=\"value\" tz=\"" + timezone + "\">" + expectedValue + "</span>";
    assertEquals(expected, html);
}
```

File: api/src/main/java/org/openmrs/module/htmlformentry/HtmlFormEntryUtil.java [hunk 5]
```diff
@@ -532,7 +516,7 @@ public static HtmlFormSchema getHtmlFormSchema(HtmlForm htmlForm, Mode mode) {
 	 * @param time the Date object that contains time information
 	 * @return a Date object with the combined date/time
 	 */
-	public static Date combineDateAndTime(Date date, Date time) {
+	public static Date combineDateAndTime(Date date, Date time, String timeZone) {
```
[the remark this answers is not in the document]
Comment (icrc-jfigueiredo) [reply api/src/main/java/org/openmrs/module/htmlformentry/HtmlFormEntryUtil.java:519]: Every place on this project that we call that function i changed the parameters to send a timezone (or null if they dont have one), we should not have a problem once they accept this changes, correct?

File: api/src/main/java/org/openmrs/module/htmlformentry/widget/TimeWidget.java (hunk 2 above)
Comment (icrc-jfigueiredo) [reply api/src/main/java/org/openmrs/module/htmlformentry/widget/TimeWidget.java:135]: Yes, correct.

File: api/src/main/java/org/openmrs/module/htmlformentry/HtmlFormEntryUtil.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (mks-d) [reply api/src/main/java/org/openmrs/module/htmlformentry/HtmlFormEntryUtil.java:519]: Other modules may rely on this API. HFE UI for instance, but even any other module that depends on HFE for whatever reason.
As a general rule, it is good practise to maintain existing APIs until they **really** need to be deprecated.

**Comment by mks-d** [comment]
@icrc-jfigueiredo you can also just force push here if there are too many changes to reconcile.
... or open a new PR, up to you.

**Pull request closed without merging**